### PR TITLE
SaveNexusPD

### LINKED ISSUE #15892: SaveDiffractionData
The file will be the histogram representation of the event data. There will be one `NXentry` with a [NXdata](http://download.nexusformat.org/doc/html/classes/base_classes/NXdata.html) for each spectrum. The algorithm should be smart enough to determine if the data is histogram or not. The tree structure will be:

* `NXentry`
  * `NXdata`
     * `data[i]` with an attribute `axes='d-spacing'`
     * `errors[i]`
     * `time-of-flight[i]` or `time-of-flight[i+1]` with an attribute `units='microsecond'`
     * `d-spacing[i]` or `d-spacing[i+1]` with an attribute `units='Angstrom'`

`GUDRUN` only supports 32-bit floats so there will need to be an option to constrain the file to using single precision.

### PR DESCRIPTION
**Pull request by peterfpeterson**
This is the first pass at a new file format that is originally geared at `GUDRUN`. There will likely be follow-on PRs as we get more feedback from Matt Tucker, Alan Soper, and others. The file format changed from the original ticket.

**To test:**

Fixes #15892.

_See the changeset for release notes_

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by peterfpeterson** [comment]
@martyngigg There is something weird with windoze on this one. It segfaults in a different place each time I try to directly run the test, but always after it has opened the file. Might upgrading `h5py` to the latest version in [thirdparty-msvc2015](https://github.com/mantidproject/thirdparty-msvc2015) fix it?

**Comment by martyngigg** [comment]
@peterfpeterson I'll try a debug build on Windows and see if that helps shed some light on it.

**Comment by martyngigg** [comment]
@peterfpeterson Can you replace the SHA1 in [buildconfig/CMake/Bootstrap.cmake](/mantidproject/mantid/blob/master/buildconfig/CMake/Bootstrap.cmake#L13)  with `de24e1f6cc307b7dfe869fef852332cfaa47be7a` to get an a rebuilt version of `h5py`? That *should* fix the segfault.

**Comment by martyngigg** [comment]
Code looks fine - methods are well named/strcutured. A unit test is in place. Moving on to functional testing

**Comment by martyngigg** [comment]
If I try setting `Compression` to either `lzw` or `None` I get an error:
```
Compression filter "lzw" is unavailable
```

or

```
Compression filter "None" is unavailable
```

Is that expected?

File: Framework/PythonInterface/plugins/algorithms/SaveNexusPD.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+                dataname = "spectrum_%d" % wksp.getSpectrum(i).getSpectrumNo()
+
+                nxdata = nxentry.create_group(dataname)
+                nxdata.attrs[self.NX_CLASS] = 'NXdata'
+
+                if self._sourcePos is None:
+                    detector = None
+                else:
+                    detector = wksp.getDetector(i)
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/SaveNexusPD.py]: I think this needs a `try..except RuntimeError:` wrapped around it, with `detector = None` for the case where `getDetector` throws. Some ISIS files have more data than has been put in the IDF, one example is `IRS38633`  in the `Data/UnitTest` directory.

**Comment by peterfpeterson** [comment]
According to h5py's website, [both lzf and gzip are available everywhere](http://docs.h5py.org/en/latest/faq.html#what-compression-processing-filters-are-supported)

File: Framework/PythonInterface/plugins/algorithms/SaveNexusPD.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+        group = 'Options'
+
+        self.declareProperty('NXentry', '',
+                             'Overrides the NXentry name from the workspace name')
+        self.declareProperty('DataType', 'float32',
+                             StringListValidator(['float32', 'float64']),
+                             doc='All data saved will be converted to this type')
+        self.declareProperty('Compression', 'gzip',
+                             StringListValidator(['gzip', 'lzw', 'None']),
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/SaveNexusPD.py]: I think this should be `StringListValidator(['gzip', 'lzf', 'None'])`. The `h5py` lists the filter as `lzf` not `lzw`.

**Comment by martyngigg** [comment]
I now get
```
Error in execution of algorithm SaveNexusPD:
'NoneType' object has no attribute 'getDistance'
  at line 295 in '/mnt/data1/source/github/mantidproject/mantid/Framework/PythonInterface/plugins/algorithms/SaveNexusPD.py'
  caused by line 179 in '/mnt/data1/source/github/mantidproject/mantid/Framework/PythonInterface/plugins/algorithms/SaveNexusPD.py'
```

for the case where `getDetector` raises an exception.

**Comment by martyngigg** [comment]
The last couple of commits have fixed the remaining issues with compression and missing detectors.

Happy to :shipit:

**Pull request merged**